File: proj.android_studio/ee-x-facebook/src/main/java/com/ee/internal/FacebookRequestDelegate.kt
package com.ee.internal

import androidx.annotation.AnyThread
import com.ee.IMessageBridge
import com.ee.Thread
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.share.widget.GameRequestDialog
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable

/**
 * Created by eps on 3/21/18.
 */
internal class FacebookRequestDelegate(
    private val _bridge: IMessageBridge,
    private val _tag: Int) : FacebookCallback<GameRequestDialog.Result> {
    companion object {
        private const val kPrefix = "FacebookRequestDelegate"
    }

    private val kOnSuccess: String
        @AnyThread get() = "${kPrefix}_onSuccess_$_tag"

    private val kOnFailure: String
        @AnyThread get() = "${kPrefix}_onFailure_$_tag"

    private val kOnCancel: String
        @AnyThread get() = "${kPrefix}_onCancel_$_tag"

    @InternalSerializationApi
    override fun onSuccess(result: GameRequestDialog.Result) {
        Thread.checkMainThread()
        @Serializable
        @Suppress("unused")
        class Response(
            val requestId: String,
            val requestRecipients: List<String>
        )

        val response = Response(result.requestId, result.requestRecipients)
        _bridge.callCpp(kOnSuccess, response.serialize())
    }

    override fun onError(error: FacebookException) {
        Thread.checkMainThread()
        _bridge.callCpp(kOnFailure, error.localizedMessage ?: "")
    }

    override fun onCancel() {
        Thread.checkMainThread()
        _bridge.callCpp(kOnCancel)
    }
}

File: src/cpp/ee/facebook/private/FacebookRequestDelegate.cpp
//
//  FacebookRequestDelegate.cpp
//  ee_x_mobile_facebook
//
//  Created by eps on 3/21/18.
//

#include "ee/facebook/private/FacebookRequestDelegate.hpp"

#include <ee/core/IMessageBridge.hpp>
#include <ee/core/Thread.hpp>
#include <ee/nlohmann/json.hpp>

namespace ee {
namespace facebook {
using Self = RequestDelegate;

namespace {
std::string k__onSuccess(int tag) {
    return "FacebookRequestDelegate_onSuccess_" + std::to_string(tag);
}

std::string k__onFailure(int tag) {
    return "FacebookRequestDelegate_onFailure_" + std::to_string(tag);
}

std::string k__onCancel(int tag) {
    return "FacebookRequestDelegate_onCancel_" + std::to_string(tag);
}
} // namespace

Self::RequestDelegate(IMessageBridge& bridge, int tag)
    : bridge_(bridge)
    , tag_(tag) {
    bridge_.registerHandler(
        [this](const std::string& message) {
            if (successCallback_) {
                auto json = nlohmann::json::parse(message);
                auto&& requestId =
                    json.value("requestId", std::string()); // Null iOS.
                auto&& requestRecipients =
                    json.value("requestRecipients",
                               std::vector<std::string>()); // Null iOS.
                successCallback_(requestId, requestRecipients);
            }
            self_.reset();
        },
        k__onSuccess(tag_));
    bridge_.registerHandler(
        [this](const std::string& message) {
            if (failureCallback_) {
                failureCallback_(message);
            }
            self_.reset();
        },
        k__onFailure(tag_));
    bridge_.registerHandler(
        [this](const std::string& message) {
            if (cancelCallback_) {
                cancelCallback_();
            }
            self_.reset();
        },
        k__onCancel(tag_));
}

Self::~RequestDelegate() {
    bridge_.deregisterHandler(k__onSuccess(tag_));
    bridge_.deregisterHandler(k__onFailure(tag_));
    bridge_.deregisterHandler(k__onCancel(tag_));
}

Self& Self::onSuccess(const SuccessCallback& callback) {
    successCallback_ = callback;
    return *this;
}

Self& Self::onFailure(const FailureCallback& callback) {
    failureCallback_ = callback;
    return *this;
}

Self& Self::onCancel(const CancelCallback& callback) {
    cancelCallback_ = callback;
    return *this;
}
} // namespace facebook
} // namespace ee
